### PR TITLE
Adds default sql indexes and table optimisations

### PR DESCRIPTION
**Pull request by Jordie0608**
This adds indexes to several tables to speed up `SELECT`, `UPDATE` and `DELETE` queries in the code. Some of the more troublesome queries don't have indexes. See this file for a listing of the queries: https://dl.dropboxusercontent.com/u/169349932/sqlqueries.xlsx

Additional most of the tables have been optimised for size and speed, where reasonable. In most cases this is changing `TEXT` to `VARCHAR(2048)`, IPs to be `UNSIGNED INT` and making a few columns `ENUM`s.

The queries are included in a separate file as they are in-fact longer than the entirety of `database_changelog.txt`

Various code changes are to ensure queries work with new data formats or restructuring for better indexes.

@MrStonedOne @nfreader
I'm certain some of the changes here will break parts of your webtools, I don't know their code to specify everything but I can prepare a list of the net differences if it'll help.

### CODE REVIEW

File: SQL/optimisations_2017-02-19.sql [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+, ADD INDEX `idx_lib_search` (`deleted` ASC, `author` ASC, `title` ASC, `category` ASC);
+
+ALTER TABLE `feedback`.`messages`
+ ADD INDEX `idx_msg_ckey_time` (`targetckey` ASC, `timestamp` ASC)
+, ADD INDEX `idx_msg_type_ckeys_time` (`type` ASC, `targetckey` ASC, `adminckey` ASC, `timestamp` ASC)
+, ADD INDEX `idx_msg_type_ckey_time_odr` (`type` ASC, `targetckey` ASC, `timestamp` ASC);
+
+ALTER TABLE `feedback`.`player`
+ ADD INDEX `idx_player_cid_ckey` (`computerid` ASC, `ckey` ASC)
+, ADD INDEX `idx_player_ip_ckey` (`ip` ASC, `ckey` ASC);
```
Comment (MrStonedOne) [inline SQL/optimisations_2017-02-19.sql]: you forgot the pre-existing one on ckey (for downstreams who don't have my modifications)

**Comment by Jordie0608** [comment]
@MrStonedOne Some things to consider
There are still more changes that could be made which I didn't do as I was either unsure about their benefit or consequences, off the top of my head:
- CIDs could be INT, I didn't think this'd really change much.
- Some VARCHAR lengths could be tweaked i.e. ckeys could be 30 instead of 32 as I think that was the max length of them.
- Some INT values could be UNSIGNED and/or SMALLINT or TINYINT, notably various ID columns will never be negatives but it's unlikely we'll be reaching the size limits for these types anytime soon.
- There's still two TEXT columns: feedback.details and library.content; these both can potentially be very long and it'd be bad if they got truncated due to length, I read a variety of conflicting statements about TEXT vs large VARCHARs as well as how InnoDB stores both either inline or not. It was rather confusing so I just left them as is to be safe.
- The three admin tables are untouched as I'd rather remove/rework them then get them working again
- I don't know all of the server's DB makeup but I'm pretty sure these queries will be quite slow, error over missing data and nuke anything not in the schema.

File: SQL/optimisations_2017-02-19.sql (hunk 1 above)
Comment (Jordie0608) [inline SQL/optimisations_2017-02-19.sql]: I don't I ever had that one (and if I did I nuked my DB when testing so it's gone either way); details? There should be a unique on ckey unless this is different.

Comment (MrStonedOne) [inline SQL/optimisations_2017-02-19.sql]: the unique is what i'm talking about.

you added it to the schema, but its not in this alter table bit.

**Comment by MrStonedOne** [comment]
>Some INT values could be UNSIGNED and/or SMALLINT or TINYINT, notably various ID columns will never be negatives but it's unlikely we'll be reaching the size limits for these types anytime soon.

unlikely to matter.

>CIDs could be INT, I didn't think this'd really change much.

IIRC the webclient fake cid can get bigger then a unsigned int, so you would have to use large unsigned int, at that point you only save 3 bytes in the avg case and lose 5 bytes in the worst case.

>There's still two TEXT columns: feedback.details and library.content; these both can potentially be very long and it'd be bad if they got truncated due to length, I read a variety of conflicting statements about TEXT vs large VARCHARs as well as how InnoDB stores both either inline or not. It was rather confusing so I just left them as is to be safe.

These should be TEXT, they are basically the only good case for doing TEXT

>The three admin tables are untouched as I'd rather remove/rework them then get them working again

Ya. thats a big mess.

>I don't know all of the server's DB makeup but I'm pretty sure these queries will be quite slow, error over missing data and nuke anything not in the schema.

I'll most likely do it when both servers are in lobby, then restart them to the new code if that was successful.

File: SQL/optimisations_2017-02-19.sql (hunk 1 above)
Comment (Jordie0608) [inline SQL/optimisations_2017-02-19.sql]: It looks like the unique key is already in the default schema as of 4 years https://github.com/tgstation/tgstation/blame/master/SQL/tgstation_schema.sql#L229
so it should be fine; although I can't say i noticed that myself ( ͡° ͜ʖ ͡°)

**Comment by MrStonedOne** [comment]
good job merging this while i'm going to be away for bit.

**Comment by optimumtact** [comment]
# _**s**_

**Comment by Jordie0608** [comment]
It'll be fine so long as nobody gets banned, unbanned, votes on a poll, dies, connects to the server with a new account or connects to the server at all.

**Pull request merged**